packages/material-ui/src/internal/SwitchBase.js

```diff
@@ -30,128 +29,122 @@ export const styles = {
+  const [checkedState, setCheckedState] = React.useState(() => {
+    if (!isControlled) {
+      return defaultChecked !== undefined ? defaultChecked : false;
     }
-  }
+    return null;
+  });
 
-  handleFocus = event => {
-    if (this.props.onFocus) {
-      this.props.onFocus(event);
+  const handleFocus = event => {
+    if (onFocus) {
+      onFocus(event);
     }
 
-    const { muiFormControl } = this.props;
     if (muiFormControl && muiFormControl.onFocus) {
       muiFormControl.onFocus(event);
     }
   };
 
-  handleBlur = event => {
-    if (this.props.onBlur) {
-      this.props.onBlur(event);
+  const handleBlur = event => {
+    if (onBlur) {
+      onBlur(event);
     }
 
-    const { muiFormControl } = this.props;
     if (muiFormControl && muiFormControl.onBlur) {
       muiFormControl.onBlur(event);
     }
   };
 
-  handleInputChange = event => {
+  const handleInputChange = event => {
     const checked = event.target.checked;
 
-    if (!this.isControlled) {
-      this.setState({ checked });
+    if (!isControlled) {
+      setCheckedState(checked);
     }
 
-    if (this.props.onChange) {
-      this.props.onChange(event, checked);
+    if (onChange) {
+      onChange(event, checked);
     }
   };
 
-  render() {
-    const {
-      autoFocus,
-      checked: checkedProp,
-      checkedIcon,
-      classes,
-      className: classNameProp,
-      defaultChecked,
-      disabled: disabledProp,
-      icon,
-      id,
-      inputProps,
-      inputRef,
-      muiFormControl,
-      name,
-      onBlur,
-      onChange,
-      onFocus,
-      readOnly,
-      required,
-      tabIndex,
-      type,
-      value,
-      ...other
-    } = this.props;
-
-    let disabled = disabledProp;
+  let disabled = disabledProp;
 
-    if (muiFormControl) {
-      if (typeof disabled === 'undefined') {
-        disabled = muiFormControl.disabled;
-      }
+  if (muiFormControl) {
+    if (typeof disabled === 'undefined') {
+      disabled = muiFormControl.disabled;
     }
+  }
 
-    const checked = this.isControlled ? checkedProp : this.state.checked;
-    const hasLabelFor = type === 'checkbox' || type === 'radio';
-
-    return (
-      <IconButton
-        component="span"
-        className={clsx(
-          classes.root,
-          {
-            [classes.checked]: checked,
-            [classes.disabled]: disabled,
-          },
-          classNameProp,
-        )}
+  const checked = isControlled ? checkedProp : checkedState;
+  const hasLabelFor = type === 'checkbox' || type === 'radio';
+
+  return (
+    <IconButton
+      component="span"
+      className={clsx(
+        classes.root,
+        {
+          [classes.checked]: checked,
+          [classes.disabled]: disabled,
+        },
+        classNameProp,
+      )}
+      disabled={disabled}
+      tabIndex={null}
+      role={undefined}
+      onFocus={handleFocus}
+      onBlur={handleBlur}
+      ref={ref}
```

Comment:
Should the ref be here? innerRef would have been spread here.

Reply:
The big question! Given we spread the props here, that we apply the `classes.root` and `className` on this element. I would say yes. cc @eps1lon.

Reply:
Sounds uncontroversial to me. 

I'm not sure what the best behavior for ref is yet. For most components it  probably make sense to only care about the outermost host (for measuring). For other components it might make more sense to attach the ref to a more useful component (e.g. focus handling, reading input value).  The latter case only applies if that "more useful component" doesn't get another props object spread. For example it might make more sense to attach the ref of `TextField` to the `input` but if we want a ref to the `input` we can already do `<Textfield InputProps={{ ref }} />`. 

I think attaching the ref to where excess props are spread i.e. `other` is the safest course of actions considering we won't need `forwardRef` in the future and it gets spread automatically.

Reply:
> I think attaching the ref to where excess props are spread i.e. other is the safest course of actions considering we won't need forwardRef in the future and it gets spread automatically.

I should have thought about this 💯 ! Yes, of course, I'm buying it 💸 .